core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,185 @@
+        tracing::info!("Finished execution of l1_batch: {l1_batch_number:?}");
+
+        metrics::histogram!(
+            "basic_witness_input_producer.input_producer_time",
+            started_at.elapsed(),
+        );
+        tracing::info!(
+            "BasicWitnessInputProducer took {:?} for L1BatchNumber {}",
+            started_at.elapsed(),
+            l1_batch_number.0
+        );
+
+        let witness_block_state = (*storage_view).borrow().witness_block_state();
+        Ok(witness_block_state)
+    }
+}
+
+#[async_trait]
+impl JobProcessor for BasicWitnessInputProducer {
+    type Job = L1BatchNumber;
+    type JobId = L1BatchNumber;
+    type JobArtifacts = WitnessBlockState;
+    const SERVICE_NAME: &'static str = "basic_witness_input_producer";
+
+    async fn get_next_job(&self) -> anyhow::Result<Option<(Self::JobId, Self::Job)>> {
+        let mut connection = self.connection_pool.access_storage().await?;
+        let l1_batch_to_process = connection
+            .basic_witness_input_producer_dal()
+            .get_next_basic_witness_input_producer_job()
+            .await?;
+        Ok(l1_batch_to_process.map(|number| (number, number)))
+    }
+
+    async fn save_failure(&self, job_id: Self::JobId, started_at: Instant, error: String) {
+        let attempts = self
+            .connection_pool
+            .access_storage()
+            .await
+            .unwrap()
+            .basic_witness_input_producer_dal()
+            .mark_job_as_failed(job_id, started_at, error)
+            .await
+            .expect("errored whilst marking job as failed");
+        match attempts {
+            Some(tries) => {
+                tracing::warn!("Failed to process job: {job_id:?}, after {tries} tries.");
+            }
+            None => {
+                tracing::warn!("L1 Batch {job_id:?} was processed successfully by another worker.");
+            }
+        }
+    }
+
+    async fn process_job(
+        &self,
+        job: Self::Job,
+        started_at: Instant,
+    ) -> JoinHandle<anyhow::Result<Self::JobArtifacts>> {
+        let validation_computational_gas_limit = self.validation_computational_gas_limit;
+        let l2_chain_id = self.l2_chain_id;
+        let connection_pool = self.connection_pool.clone();
+        tokio::task::spawn_blocking(move || {
+            let rt_handle = Handle::current();
+            Self::process_job_impl(
+                rt_handle,
+                job,
+                started_at,
+                connection_pool.clone(),
+                validation_computational_gas_limit,
+                l2_chain_id,
+            )
+        })
+    }
+
+    async fn save_result(
+        &self,
+        job_id: Self::JobId,
+        started_at: Instant,
+        artifacts: Self::JobArtifacts,
+    ) -> anyhow::Result<()> {
+        let upload_started_at = Instant::now();
+        let object_path = self.object_store.put(job_id, &artifacts).await?;
+        metrics::histogram!(
```

Comment:
Please don't forget to migrate the module to `vise`. I failed to find a task in linear for that, probably it makes sense to create one, e.g. as a sub-task in [the umbrella issue](https://linear.app/matterlabs/issue/PLA-166/witness-generation-remove-dependency-from-db).